oximeter/instruments/src/kstat/link.rs

```diff
@@ -740,4 +740,97 @@ mod tests {
         // kept the final samples.
         assert!(count.value() > 4096);
     }
+
+    // Regression for https://github.com/oxidecomputer/omicron/issues/8889
+    #[tokio::test]
+    async fn updating_target_does_not_duplicate_samples() {
+        let log = test_logger();
+        let sampler = KstatSampler::new(&log).unwrap();
+        let link = TestEtherstub::new();
+        info!(log, "created test etherstub"; "name" => &link.name);
+        let target = SledDataLinkTarget {
+            rack_id: RACK_ID,
+            sled_id: SLED_ID,
+            sled_serial: SLED_SERIAL.into(),
+            link_name: link.name.clone().into(),
+            kind: KIND.into(),
+            sled_model: SLED_MODEL.into(),
+            sled_revision: SLED_REVISION,
+            zone_name: ZONE_NAME.into(),
+        };
+        let dl = SledDataLink::new(target.clone(), true);
+        let collection_interval = Duration::from_millis(10);
+        let details = CollectionDetails::never(collection_interval);
+        let _id = sampler.add_target(dl.clone(), details).await.unwrap();
+
+        // Update the target.
+        sampler
+            .update_target(
+                dl,
+                CollectionDetails::never(Duration::from_millis(15)),
+            )
+            .await
+            .unwrap();
+
+        // We'll wait for 150ms. That would be 15 collections at the original
+        // interval, but 10 at the new one. So we can assert that we've actually
+        // updated things, and certainly don't have 25!.
+        tokio::time::pause();
+        const MAX_DURATION: Duration = Duration::from_millis(150);
+        const STEP_DURATION: Duration = Duration::from_millis(1);
+        let now = Instant::now();
+        while now.elapsed() < MAX_DURATION {
+            tokio::time::advance(STEP_DURATION).await;
+        }
+
+        // The assertion here is pretty weak, but we want to make sure that we
+        // don't have double the expected number of collections.
```

Comment:
Is this to make the test less flaky even though we're using `tokio::time::advance` instead of sleeping? If so, is it still possible this might flake (e.g., on a very overloaded CI runner)?

It would be a weirdly specific test, but I wonder if we should instead check the futures unordered set itself and confirm:

* there are no duplicate IDs
* the interval for this target is 15ms, not 10ms

which shouldn't have any flake potential at all.

Reply:
Yeah, this is to be "strict enough" to test something, without flakes. It's not great. This code doesn't have access to the actual set of futures, because that's in a separate task entirely. I could add yet another message type for asking that task about the set of futures it's awaiting. I'm on the fence about that, but it would definitely avoid flakes.

Reply:
I've done the test-only messages for other bits. It's not too terrible, and would be much better. Let me try that and see how it goes. We can always reduce flakiness by making the difference between the original and updated sampling intervals bigger, and advancing time further -- the expected difference grows in that case, so we're less likely to have false-positives.